src/internal/m365/collection/drive/collections.go

```diff
@@ -257,76 +257,78 @@ func (c *Collections) Get(
+		driveIDToDeltaLink = map[string]string{}
+		driveIDToPrevPaths = map[string]map[string]string{}
+		numPrevItems       = 0
 	)
 
 	for _, d := range drives {
 		var (
-			driveID     = ptr.Val(d.GetId())
-			driveName   = ptr.Val(d.GetName())
-			prevDelta   = prevDeltas[driveID]
-			oldPaths    = oldPathsByDriveID[driveID]
-			numOldDelta = 0
-			ictx        = clues.Add(ctx, "drive_id", driveID, "drive_name", driveName)
+			driveID   = ptr.Val(d.GetId())
+			driveName = ptr.Val(d.GetName())
+			ictx      = clues.Add(ctx, "drive_id", driveID, "drive_name", driveName)
+
+			excludedItemIDs = map[string]struct{}{}
+			oldPrevPaths    = oldPrevPathsByDriveID[driveID]
+			prevDeltaLink   = prevDriveIDToDelta[driveID]
 		)
 
 		delete(driveTombstones, driveID)
 
-		if _, ok := c.CollectionMap[driveID]; !ok {
-			c.CollectionMap[driveID] = map[string]*Collection{}
+		if _, ok := driveIDToPrevPaths[driveID]; !ok {
+			driveIDToPrevPaths[driveID] = map[string]string{}
 		}
 
-		if len(prevDelta) > 0 {
-			numOldDelta++
+		if _, ok := c.CollectionMap[driveID]; !ok {
+			c.CollectionMap[driveID] = map[string]*Collection{}
 		}
 
 		logger.Ctx(ictx).Infow(
 			"previous metadata for drive",
-			"num_paths_entries", len(oldPaths),
-			"num_deltas_entries", numOldDelta)
+			"num_paths_entries", len(oldPrevPaths))
 
-		delta, paths, excluded, err := collectItems(
+		items, du, err := c.handler.EnumerateDriveItemsDelta(
 			ictx,
-			c.handler.NewItemPager(driveID, "", api.DriveItemSelectDefault()),
 			driveID,
-			driveName,
-			c.UpdateCollections,
-			oldPaths,
-			prevDelta,
-			errs)
+			prevDeltaLink)
 		if err != nil {
 			return nil, false, err
 		}
 
-		// Used for logging below.
-		numDeltas := 0
-
 		// It's alright to have an empty folders map (i.e. no folders found) but not
 		// an empty delta token. This is because when deserializing the metadata we
 		// remove entries for which there is no corresponding delta token/folder. If
 		// we leave empty delta tokens then we may end up setting the State field
 		// for collections when not actually getting delta results.
-		if len(delta.URL) > 0 {
-			deltaURLs[driveID] = delta.URL
-			numDeltas++
+		if len(du.URL) > 0 {
+			driveIDToDeltaLink[driveID] = du.URL
+		}
+
+		newPrevPaths, err := c.UpdateCollections(
+			ctx,
+			driveID,
+			driveName,
+			items,
+			oldPrevPaths,
+			driveIDToPrevPaths[driveID],
+			excludedItemIDs,
+			du.Reset,
+			errs)
+		if err != nil {
+			return nil, false, clues.Stack(err)
 		}
 
 		// Avoid the edge case where there's no paths but we do have a valid delta
 		// token. We can accomplish this by adding an empty paths map for this
 		// drive. If we don't have this then the next backup won't use the delta
 		// token because it thinks the folder paths weren't persisted.
-		folderPaths[driveID] = map[string]string{}
-		maps.Copy(folderPaths[driveID], paths)
+		driveIDToPrevPaths[driveID] = map[string]string{}
+		maps.Copy(driveIDToPrevPaths[driveID], newPrevPaths)
```

Comment:
> The usage of this variable is extremely confusing and difficult to track. It's also not clear that it results in correct behavior under some edge cases

Oh, believe me, I am 100% there with you.  That said, these requests are a bit out of scope of this change.  The goal was only to refactor the drive enumeration, and to make the minimum amount of logic changes to the code calling it.  I think I've been able to do that, and am certainly open to the possibility that I messed something up because, like you said, this is complicated as all get-out.

As for a sanity refactor to this code itself, that's another task, and I've already completely overblown my timebox on this change to begin with.  I'll create a tech debt ticket for it, but would prefer to keep the work to that space.

Reply:
issue: https://github.com/alcionai/corso/issues/4379

Reply:
Can we at least go back to not reusing the same variable for different purposes? IIRC, the original code used the per-drive item set for detecting deleted items. I'm fine with addressing some of this in a follow up refactor, but I'm not comfortable with making the code more difficult to read by reusing a variable like this

Reply:
sure, can do